docs: initial `metavariable-name.module{,s}` (#2056)
semgrep/semgrep-docs: docs/writing-rules/rule-syntax.md

### `metavariable-name`

The `metavariable-name` operator adds a constraint to the types of identifiers a metavariable is able to match. Currently the only constraint supported is on module or namespace an identifier originates from. This is useful for filtering results in languages which don't have a native syntax for fully qualified names, or languages where module names may contain characters which are not legal in identifiers (e.g., JavaScript, TypeScript). 

:::info
The `metavariable-name` operator requires the Pro engine, since it requires additional name resolution information. Moreover, name resolution information is higher fidelity with interfile enabled, and thus `metavariable-name` will be more accurate with interfile naming.
:::

```yaml
rules:
  - id: insecure-method
    patterns:
      - pattern: $MODULE.method(...)
      - metavariable-name:
          metavariable: $MODULE
          module: "@foo-bar"
    message: Uses insecure method from @foo-bar.
    languages:
      - javascript
    severity: ERROR
```

The pattern immediately above matches the following:

```javascript

> import * as lib from '@foo-bar';

Reviewer: I would show both import and require in the example, just to show that it works for both
```javascript
// ECMAScript modules
import * as lib from '@foo-bar'
import * as lib2 from 'myotherlib'

// CommonJS modules
const { method } = require('@foo-bar')
const lib3 = require('myotherlib')

// ruleid: insecure-method
lib.method("test")

// ruleid: insecure-method
method("test")

// ok: insecure-method
lib.method2("test")

// ok: insecure-method
lib2.method("test")

// ok: insecure-method
lib3.method("test")
```

This removes quotes (`'`, `"`, and `` ` ``) from both ends of the metavariable content. As a result, Semgrep detects `"2147483648"`, but it does **not** detect `"2147483646"`. This is useful when you expect strings to contain integer or float data.

### `metavariable-name`

> The `metavariable-name` operator adds a constraint to the types of identifiers a metavariable is able to match. Currently the only constraint supported is on module or namespace an identifier originates from. This is useful for filtering results in languages which don't have a native syntax for fully qualified names, or languages where module names may contain characters which are not legal in identifiers (e.g., JavaScript, TypeScript). 

Reviewer 2: ```suggestion
The `metavariable-name` operator adds a constraint to the types of identifiers a metavariable is able to match. Currently the only constraint supported is on the **module** or **namespace** an identifier originates from. This is useful for filtering results in languages which don't have a native syntax for fully qualified names, or languages where module names may contain characters which are not legal in identifiers, such as JavaScript or TypeScript). 
```

Reviewer 2: @kopecs This is the only remaining comment to resolve - are you ok with committing the suggestion?

Author: Bold/italics on module or namespace seem a bit gratuitous to me, but adding "the" and rewording to "such as" instead of "e.g." sounds good.

:::info

> The `metavariable-name` operator requires the Pro engine, since it requires additional name resolution information. Moreover, name resolution information is higher fidelity with interfile enabled, and thus `metavariable-name` will be more accurate with interfile naming.

Reviewer 2: ```suggestion
The `metavariable-name` operator requires cross-file analysis, which you can enable following the steps in [Enable cross-file analysis](/docs/getting-started/cli#enable-cross-file-analysis). This is because it requires additional name resolution information. Moreover, name resolution information is more precise with cross-file enabled, and thus `metavariable-name` will be more accurate with cross-file naming.
```

Author: It does require Pro but not interfile (but works better with interfile). Do you have a preferred wording for the first sentence, given that?

Reviewer 2: @kopecs in the **Semgrep AppSec Platform > Settings** tab, there is no setting in which we enable "only" Pro - the UI toggle enables both. Also, the toggle explicitly calls it **Cross-file** 😅 so I can't change the toggle name. 

Maybe we can say:

> The `metavariable-name` operator requires _advanced_ analysis, which you can enable following the steps in [Enable cross-file analysis](/docs/getting-started/cli#enable-cross-file-analysis) ...

Author: OK. Let me confirm what the default here is if that toggle is off. Not sure if we still have Pro intrafile or if it is OSS (or it depends on some other thing). We should disclaim it here either way, since CE users won't have access to it, but let me get back to you.

Reviewer 2: Anything "Pro" branded is by default not OSS, so yeah, CE users will not have access to it. Are you ok with me pushing a commit to add a callout?

Author: Feel free to push whatever!

This removes quotes (`'`, `"`, and `` ` ``) from both ends of the metavariable content. As a result, Semgrep detects `"2147483648"`, but it does **not** detect `"2147483646"`. This is useful when you expect strings to contain integer or float data.

### `metavariable-name`

:::info

> `metavariable-name` requires a Semgrep account and **[cross-file analysis](/docs/getting-started/cli#enable-cross-file-analysis)**. 

Author: Seems redundant with the updated version at lines 771-3? Cross-file isn't required though, so if we want to keep this it should either just say Semgrep account or Pro engine.

Author: You do not _need_ cross-file. But, the results are better with cross-file enabled.

The pro engine is a requirement. This is not quite the same as being logged-in, since you can pass, e.g., `--oss-only`.
